cli/config_file.go

```diff
@@ -384,3 +389,93 @@ func RandomChars(bufflength uint) string {
 
 	return result[0:bufflength]
 }
+
+const (
+	CruxAgentGrpcPort  = "CruxAgentGrpcPort"
+	CruxGrpcPort       = "CruxGrpcPort"
+	CruxUIPort         = "CruxUIPort"
+	KratosAdminPort    = "KratosAdminPort"
+	KratosPublicPort   = "KratosPublicPort"
+	KratosPostgresPort = "KratosPostgresPort"
+	MailSlurperPort    = "MailSlurperPort"
+	MailSlurperPort2   = "MailSlurperPort2"
+	CruxPostgresPort   = "CruxPostgresPort"
+)
+
+func CheckAndUpdatePorts(settings *Settings) *Settings {
+	portMap := map[string]uint{}
+	portMap[CruxAgentGrpcPort] = getAvailablePort(portMap, settings.SettingsFile.Options.CruxAgentGrpcPort, CruxAgentGrpcPort)
+	portMap[CruxGrpcPort] = getAvailablePort(portMap, settings.SettingsFile.Options.CruxGrpcPort, CruxGrpcPort)
+	portMap[CruxUIPort] = getAvailablePort(portMap, settings.SettingsFile.Options.CruxUIPort, CruxUIPort)
+	portMap[CruxPostgresPort] = getAvailablePort(portMap, settings.SettingsFile.Options.CruxPostgresPort, CruxPostgresPort)
+	portMap[KratosAdminPort] = getAvailablePort(portMap, settings.SettingsFile.Options.KratosAdminPort, KratosAdminPort)
+	portMap[KratosPublicPort] = getAvailablePort(portMap, settings.SettingsFile.Options.KratosPublicPort, KratosPublicPort)
+	portMap[KratosPostgresPort] = getAvailablePort(portMap, settings.SettingsFile.Options.KratosPostgresPort, KratosPostgresPort)
+	portMap[MailSlurperPort] = getAvailablePort(portMap, settings.SettingsFile.Options.MailSlurperPort, MailSlurperPort)
+	portMap[MailSlurperPort2] = getAvailablePort(portMap, settings.SettingsFile.Options.MailSlurperPort2, MailSlurperPort2)
+
+	settings.SettingsFile.Options.CruxAgentGrpcPort = portMap[CruxAgentGrpcPort]
+	settings.SettingsFile.Options.CruxGrpcPort = portMap[CruxGrpcPort]
+	settings.SettingsFile.Options.CruxUIPort = portMap[CruxUIPort]
+	settings.SettingsFile.Options.CruxPostgresPort = portMap[CruxPostgresPort]
+	settings.SettingsFile.Options.KratosAdminPort = portMap[KratosAdminPort]
+	settings.SettingsFile.Options.KratosPublicPort = portMap[KratosPublicPort]
+	settings.SettingsFile.Options.KratosPostgresPort = portMap[KratosPostgresPort]
+	settings.SettingsFile.Options.MailSlurperPort = portMap[MailSlurperPort]
+	settings.SettingsFile.Options.MailSlurperPort2 = portMap[MailSlurperPort2]
+
+	return settings
+}
+
+func getAvailablePort(portMap map[string]uint, portNum uint, portDesc string) uint {
+	for {
+		if err := portIsAvailable(portMap, portNum); err != nil {
+			fmt.Fprintf(os.Stderr, "error in binding port for %s: %s\n", portDesc, err.Error())
+			fmt.Fprintf(os.Stdout, "type another port: ")
+			portNum = scanPort(portNum)
+			continue
+		}
+		break
+	}
+	return portNum
+}
+
+func scanPort(portNum uint) uint {
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		newPort, err := strconv.ParseUint(scanner.Text(), 10, 32)
```

Comment:
These are considered magic numbers, please make these a const.

Reply:
If you run that tool in the cli directory it should work as is (there wasn't any error I'm aware of). The configuration file can be found in the project root.